File: KrakTram/TypeFavour.vb
Imports System.Xml.Serialization

<XmlType("stop")>
Public Class FavStop
    <XmlAttribute()>
    Public Property Lat As Double
    <XmlAttribute()>
    Public Property Lon As Double
    <XmlAttribute()>
    Public Property Name As String
    <XmlAttribute()>
    Public Property maxOdl As Integer
End Class

Public Class FavStopList

    Private Itemy As Collection(Of FavStop) = New Collection(Of FavStop)
    Private bDirty As Boolean = False

    Public Sub Add(sName As String, dLat As Double, dLon As Double, iMaxOld As Integer)

        ' jesli jest (np. tram), to nie dodawaj (np. bus)
        For Each oNode As FavStop In Itemy
            If oNode.Name = sName Then Exit Sub
        Next

        ' nie ma, to dodaj
        Dim oNew As FavStop = New FavStop
        oNew.Lat = dLat
        oNew.Lon = dLon
        oNew.Name = sName
        oNew.maxOdl = iMaxOld
        Itemy.Add(oNew)
        bDirty = True
    End Sub

    Public Sub Del(sName As String)
        For Each oItem As FavStop In Itemy
            If oItem.Name = sName Then
                Itemy.Remove(oItem)
                Exit Sub
            End If
        Next
    End Sub

    Public Sub InitPkar()
        Add("domek", 50.0198, 19.9785, 500)
        Add("meiselsa", 50.0513642, 19.9432361, 600)
        Add("franc3", 50.059781, 19.9339632, 500)
        Add("widok", 50.0789713, 19.8816113, 500)
        Save(False)
        App.SetSettingsBool("pkarmode", True, True) ' roaming data
    End Sub

    ' Load
    Private Async Function Load() As Task(Of Boolean)
        ' ret=false gdy nie jest wczytane

        Dim oObj As Windows.Storage.StorageFile =
            Await Windows.Storage.ApplicationData.Current.LocalCacheFolder.TryGetItemAsync("favs.xml")
        If oObj Is Nothing Then Return False
        Dim oFile As Windows.Storage.StorageFile = TryCast(oObj, Windows.Storage.StorageFile)

        Dim oSer As XmlSerializer = New XmlSerializer(GetType(Collection(Of FavStop)))
        Dim oStream As Stream = Await oFile.OpenStreamForReadAsync
        Itemy = TryCast(oSer.Deserialize(oStream), Collection(Of FavStop))
        bDirty = False
        Return True
    End Function

    Public Async Function Save(bForce As Boolean) As Task

        If Not bForce And Not bDirty Then Exit Function

        Dim oFile As Windows.Storage.StorageFile =
            Await Windows.Storage.ApplicationData.Current.LocalCacheFolder.CreateFileAsync(
                "favs.xml", Windows.Storage.CreationCollisionOption.ReplaceExisting)

        If oFile Is Nothing Then Exit Function

        Dim oSer As XmlSerializer = New XmlSerializer(GetType(Collection(Of FavStop)))
        Dim oStream As Stream = Await oFile.OpenStreamForWriteAsync
        oSer.Serialize(oStream, Itemy)
        oStream.Dispose()   ' == fclose
        bDirty = False
    End Function

    Private Async Function Import() As Task(Of Boolean)
        Dim sOldVers As String = App.GetSettingsString("favPlaces")
        If sOldVers.Length < 25 Then    ' "<places></places>
            App.SetSettingsString("favPlaces", "")
            Return False
        End If

        ' skoro jest zmienna, to nalezy to zaimportowac
        Dim bError As Boolean = False
        Dim oXmlPlaces As New Windows.Data.Xml.Dom.XmlDocument
        Try
            oXmlPlaces.LoadXml(sOldVers)
        Catch ex As Exception
            bError = True
        End Try
        If bError Then
            App.DialogBox("ERROR loading favourites list")
            Return False
        End If

        For Each oPlace As Windows.Data.Xml.Dom.IXmlNode In oXmlPlaces.DocumentElement.SelectNodes("//place")
            Add(oPlace.SelectSingleNode("@name").NodeValue,
                oPlace.SelectSingleNode("@lat").NodeValue,
                oPlace.SelectSingleNode("@long").NodeValue,
                oPlace.SelectSingleNode("@maxOdl").NodeValue)
        Next

        Await Save(True)
        App.SetSettingsString("favPlaces", "")
        App.DialogBoxRes("resImportedOldFav")
        Return True
    End Function


    Public Async Function LoadOrImport() As Task

        If Await Load() Then Exit Function
        Import()

    End Function

    Public Function GetList() As ICollection(Of FavStop)
        Return Itemy
    End Function
End Class

File: KrakTram_Uno/KrakTram_Uno.Shared/TypeFavour.cs
using System;
using System.IO;
using System.Xml.Serialization;
using vb14 = VBlib.pkarlibmodule14;

// Kolejne wersje:
// GetSettingsString("favPlaces") i w nim XML
// LocalCacheFolder("favs.xml"), dalej XML
// 2022.01.31, LocalFolder("favs.json") plus import z pliku XML oraz ze zmiennej (WinSTORE, nie ma AntroSTORE!)
// 2022.04.xx, zmiana app wymuszona przez podział przystanku na słupki - 

// pozostałość po poprzednim formacie (XML), potrzebne dla Import

namespace KrakTram

{

    [XmlType("stop")]
    public class FavStopXML
    {
        [XmlAttribute()]
        public double Lat { get; set; }
        [XmlAttribute()]
        public double Lon { get; set; }
        [XmlAttribute()]
        public string Name { get; set; }
        [XmlAttribute()]
        public int maxOdl { get; set; }
    }

    /// <summary>Tylko na potrzeby migracji danych do JSON i VBLib</summary>

    public class FavStopListXML
    {
        private static System.Collections.ObjectModel.Collection<FavStopXML> Itemy = new System.Collections.ObjectModel.Collection<FavStopXML>();

        public static async System.Threading.Tasks.Task<bool> Load()
        {
            // ret=false gdy nie jest wczytane

            Windows.Storage.StorageFile oObj = (await Windows.Storage.ApplicationData.Current.LocalCacheFolder.TryGetItemAsync("favs.xml")) as Windows.Storage.StorageFile;
            if (oObj == null)
                return false;
            Windows.Storage.StorageFile oFile = oObj as Windows.Storage.StorageFile;

            try
            {
                XmlSerializer oSer = new XmlSerializer(typeof(System.Collections.ObjectModel.Collection<FavStopXML>));
                Stream oStream = await oFile.OpenStreamForReadAsync();
                System.Xml.XmlReader oXmlReader = System.Xml.XmlReader.Create(oStream);
                Itemy = oSer.Deserialize(oXmlReader) as System.Collections.ObjectModel.Collection<FavStopXML>;
                oXmlReader.Dispose();
            }
            catch
            {
                vb14.DialogBox("ERROR reading fav file?");
                return false;
            }

            foreach (var oXmlItem in Itemy)
            {
                App.oFavour.Add(oXmlItem.Name, oXmlItem.Lat, oXmlItem.Lon, oXmlItem.maxOdl);
            }
            return true;
        }

#if false
        public static async System.Threading.Tasks.Task<bool> Import()
        {
            string sOldVers = vb14.GetSettingsString("favPlaces");
            if (sOldVers.Length < 25)
            {
                vb14.SetSettingsString("favPlaces", "");
                return false;
            }

#if NETFX_CORE
            // skoro jest zmienna, to nalezy to zaimportowac - ale moze byc tylko na Windows...
            bool bError = false;
            var oXmlPlaces = new Windows.Data.Xml.Dom.XmlDocument();
            try
            {
                oXmlPlaces.LoadXml(sOldVers);
            }
            catch
            {
                bError = true;
            }
            if (bError)
            {
                await VBlib.pk.DialogBoxAsync("ERROR loading favourites list");
                return false;
            }

            foreach (Windows.Data.Xml.Dom.IXmlNode oPlace in oXmlPlaces.DocumentElement.SelectNodes("//place"))
            {
                App.oFavour.Add(oPlace.SelectSingleNode("@name").NodeValue.ToString(), (double)oPlace.SelectSingleNode("@lat").NodeValue, (double)oPlace.SelectSingleNode("@long").NodeValue, (int)oPlace.SelectSingleNode("@maxOdl").NodeValue);
            }

            VBlib.pk.SetSettingsString("favPlaces", "");
            await VBlib.pk.DialogBoxResAsync("resImportedOldFav");
            return true;
#else
            return false;   // ale to się nie zdarzy - bo wtedy nie może być OldFav :)
#endif
        }
#endif 


    }

}
